pkg/lightspeed/funcs.go

```diff
@@ -0,0 +1,237 @@
+	OLSConfigName = "cluster"
+)
+
+// GetOLSConfig returns OLSConfig if there is one present in the cluster.
+func GetOLSConfig(ctx context.Context, helper *common_helper.Helper) (uns.Unstructured, error) {
+	OLSConfigGVR := schema.GroupVersionResource{
+		Group:    "ols.openshift.io",
+		Version:  "v1alpha1",
+		Resource: "olsconfigs",
+	}
+
+	OLSConfigList := &uns.UnstructuredList{}
+	OLSConfigList.SetGroupVersionKind(OLSConfigGVR.GroupVersion().WithKind("OLSConfig"))
+	err := helper.GetClient().List(ctx, OLSConfigList)
+	if err != nil {
+		return uns.Unstructured{}, err
+	}
+
+	if OLSConfigList.Items != nil && len(OLSConfigList.Items) > 0 {
+		return OLSConfigList.Items[0], nil
+	}
+
+	return uns.Unstructured{}, k8s_errors.NewNotFound(
+		schema.GroupResource{Group: "ols.openshifg.io", Resource: "olsconfigs"},
+		"OLSConfig")
+}
+
+// IsOLSOperatorInstalled checks whether OLS Operator is already running in the cluster.
+func IsOLSOperatorInstalled(ctx context.Context, helper *common_helper.Helper) (bool, error) {
+	csvGVR := schema.GroupVersionResource{
+		Group:    "operators.coreos.com",
+		Version:  "v1alpha1",
+		Resource: "clusterserviceversions",
+	}
+
+	csvList := &uns.UnstructuredList{}
+	csvList.SetGroupVersionKind(csvGVR.GroupVersion().WithKind("clusterserviceversion"))
+	err := helper.GetClient().List(ctx, csvList)
+	if err != nil {
+		return false, err
+	}
+
+	for _, csv := range csvList.Items {
+		if strings.HasPrefix(csv.GetName(), "lightspeed-operator") {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// PatchOLSConfig patches OLSConfig with information from OpenStackLightspeed instance.
+func PatchOLSConfig(
+	olsConfig *uns.Unstructured,
+	instance *lightspeedv1beta1.OpenStackLightspeed,
+	helper *common_helper.Helper,
+) error {
+	// 1. Patch the Providers section
+	providersPatch := []interface{}{
+		map[string]interface{}{
+			"credentialsSecretRef": map[string]interface{}{
+				"name": StringPtrValue(instance.Spec.LLMCredentials),
+			},
+			"models": []interface{}{
+				map[string]interface{}{
+					"name":       StringPtrValue(instance.Spec.ModelName),
+					"parameters": map[string]interface{}{},
+				},
+			},
+			"name": OpenStackLightspeedDefaultProvider,
+			"type": StringPtrValue(instance.Spec.LLMEndpointType),
+			"url":  StringPtrValue(instance.Spec.LLMEndpoint),
+		},
+	}
+
+	if err := uns.SetNestedSlice(olsConfig.Object, providersPatch, "spec", "llm", "providers"); err != nil {
+		return err
+	}
+
+	indexID, err := getIndexID(instance.Spec.RAGImage)
+	if err != nil {
+		return err
+	}
+
+	// 2. Patch the RAG section
+	openstackRAG := []interface{}{
+		map[string]interface{}{
+			"image":     instance.Spec.RAGImage,
+			"indexID":   indexID,
+			"indexPath": OpenStackLightspeedVectorDBPath,
+		},
+	}
+
+	if err := uns.SetNestedSlice(olsConfig.Object, openstackRAG, "spec", "ols", "rag"); err != nil {
+		return err
+	}
+
+	tlsCaCertBundle := StringPtrValue(instance.Spec.TLSCACertBundle)
```

Comment:
same questions as in the api file, if this should be the CA bundle from the ctlplane instead.